File: multi_agent/projects/im-mobile/lib/models/group_announcement.dart
/**
 * 群公告数据模型 - Flutter移动端
 * 
 * @author IM System
 * @version 1.0
 */

import 'dart:convert';

/// 群公告模型
class GroupAnnouncement {
  final String announcementId;
  final String groupId;
  final String authorId;
  final String authorName;
  final String? authorAvatar;
  final String title;
  final String content;
  final String? summary;
  final bool pinned;
  final bool edited;
  final DateTime? editedAt;
  final DateTime createdAt;
  final DateTime updatedAt;
  final int readCount;
  final int unreadCount;
  final int commentCount;

  GroupAnnouncement({
    required this.announcementId,
    required this.groupId,
    required this.authorId,
    required this.authorName,
    this.authorAvatar,
    required this.title,
    required this.content,
    this.summary,
    this.pinned = false,
    this.edited = false,
    this.editedAt,
    required this.createdAt,
    required this.updatedAt,
    this.readCount = 0,
    this.unreadCount = 0,
    this.commentCount = 0,
  });

  /// 从JSON创建
  factory GroupAnnouncement.fromJson(Map<String, dynamic> json) {
    return GroupAnnouncement(
      announcementId: json['announcementId'] as String,
      groupId: json['groupId'] as String,
      authorId: json['authorId'] as String,
      authorName: json['authorName'] as String,
      authorAvatar: json['authorAvatar'] as String?,
      title: json['title'] as String,
      content: json['content'] as String,
      summary: json['summary'] as String?,
      pinned: json['pinned'] as bool? ?? false,
      edited: json['edited'] as bool? ?? false,
      editedAt: json['editedAt'] != null
          ? DateTime.parse(json['editedAt'] as String)
          : null,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
      readCount: json['readCount'] as int? ?? 0,
      unreadCount: json['unreadCount'] as int? ?? 0,
      commentCount: json['commentCount'] as int? ?? 0,
    );
  }

  /// 转换为JSON
  Map<String, dynamic> toJson() {
    return {
      'announcementId': announcementId,
      'groupId': groupId,
      'authorId': authorId,
      'authorName': authorName,
      'authorAvatar': authorAvatar,
      'title': title,
      'content': content,
      'summary': summary,
      'pinned': pinned,
      'edited': edited,
      'editedAt': editedAt?.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'readCount': readCount,
      'unreadCount': unreadCount,
      'commentCount': commentCount,
    };
  }

  /// 复制并修改
  GroupAnnouncement copyWith({
    String? announcementId,
    String? groupId,
    String? authorId,
    String? authorName,
    String? authorAvatar,
    String? title,
    String? content,
    String? summary,
    bool? pinned,
    bool? edited,
    DateTime? editedAt,
    DateTime? createdAt,
    DateTime? updatedAt,
    int? readCount,
    int? unreadCount,
    int? commentCount,
  }) {
    return GroupAnnouncement(
      announcementId: announcementId ?? this.announcementId,
      groupId: groupId ?? this.groupId,
      authorId: authorId ?? this.authorId,
      authorName: authorName ?? this.authorName,
      authorAvatar: authorAvatar ?? this.authorAvatar,
      title: title ?? this.title,
      content: content ?? this.content,
      summary: summary ?? this.summary,
      pinned: pinned ?? this.pinned,
      edited: edited ?? this.edited,
      editedAt: editedAt ?? this.editedAt,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      readCount: readCount ?? this.readCount,
      unreadCount: unreadCount ?? this.unreadCount,
      commentCount: commentCount ?? this.commentCount,
    );
  }

  @override
  String toString() {
    return 'GroupAnnouncement(announcementId: $announcementId, title: $title, pinned: $pinned)';
  }
}

/// 公告已读状态
class AnnouncementReadStatus {
  final String statusId;
  final String announcementId;
  final String userId;
  final bool read;
  final DateTime? readAt;

  AnnouncementReadStatus({
    required this.statusId,
    required this.announcementId,
    required this.userId,
    this.read = false,
    this.readAt,
  });

  factory AnnouncementReadStatus.fromJson(Map<String, dynamic> json) {
    return AnnouncementReadStatus(
      statusId: json['statusId'] as String,
      announcementId: json['announcementId'] as String,
      userId: json['userId'] as String,
      read: json['read'] as bool? ?? false,
      readAt: json['readAt'] != null
          ? DateTime.parse(json['readAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'statusId': statusId,
      'announcementId': announcementId,
      'userId': userId,
      'read': read,
      'readAt': readAt?.toIso8601String(),
    };
  }
}

/// 创建公告请求
class CreateAnnouncementRequest {
  final String groupId;
  final String authorId;
  final String authorName;
  final String? authorAvatar;
  final String title;
  final String content;

  CreateAnnouncementRequest({
    required this.groupId,
    required this.authorId,
    required this.authorName,
    this.authorAvatar,
    required this.title,
    required this.content,
  });

  Map<String, dynamic> toJson() {
    return {
      'groupId': groupId,
      'authorId': authorId,
      'authorName': authorName,
      'authorAvatar': authorAvatar,
      'title': title,
      'content': content,
    };
  }
}

/// 更新公告请求
class UpdateAnnouncementRequest {
  final String userId;
  final String title;
  final String content;

  UpdateAnnouncementRequest({
    required this.userId,
    required this.title,
    required this.content,
  });

  Map<String, dynamic> toJson() {
    return {
      'userId': userId,
      'title': title,
      'content': content,
    };
  }
}

/// 批量标记已读请求
class BatchMarkReadRequest {
  final String groupId;
  final String userId;
  final List<String> announcementIds;

  BatchMarkReadRequest({
    required this.groupId,
    required this.userId,
    required this.announcementIds,
  });

  Map<String, dynamic> toJson() {
    return {
      'groupId': groupId,
      'userId': userId,
      'announcementIds': announcementIds,
    };
  }
}

/// 公告列表响应
class AnnouncementListResponse {
  final List<GroupAnnouncement> announcements;
  final int total;
  final int page;
  final int size;

  AnnouncementListResponse({
    required this.announcements,
    required this.total,
    required this.page,
    required this.size,
  });

  factory AnnouncementListResponse.fromJson(Map<String, dynamic> json) {
    return AnnouncementListResponse(
      announcements: (json['announcements'] as List<dynamic>)
          .map((e) => GroupAnnouncement.fromJson(e as Map<String, dynamic>))
          .toList(),
      total: json['total'] as int,
      page: json['page'] as int,
      size: json['size'] as int,
    );
  }
}

/// 公告已读统计
class AnnouncementReadStats {
  final String announcementId;
  final int readCount;
  final int unreadCount;
  final int totalMembers;

  AnnouncementReadStats({
    required this.announcementId,
    required this.readCount,
    required this.unreadCount,
    required this.totalMembers,
  });

  factory AnnouncementReadStats.fromJson(Map<String, dynamic> json) {
    return AnnouncementReadStats(
      announcementId: json['announcementId'] as String,
      readCount: json['readCount'] as int,
      unreadCount: json['unreadCount'] as int,
      totalMembers: json['totalMembers'] as int,
    );
  }
}

File: im-backend/src/main/java/com/im/server/entity/GroupAnnouncement.java
package com.im.server.entity;

import java.time.LocalDateTime;
import java.util.Objects;

/**
 * 群公告实体类
 * 
 * @author IM System
 * @version 1.0
 */
public class GroupAnnouncement {

    private String announcementId;        // 公告ID
    private String groupId;                // 群ID
    private String authorId;               // 发布者ID
    private String authorName;             // 发布者昵称
    private String authorAvatar;            // 发布者头像
    private String title;                  // 公告标题
    private String content;                // 公告内容（支持Markdown）
    private String summary;                // 内容摘要（用于预览）
    private boolean pinned;                // 是否置顶
    private boolean edited;                // 是否被编辑过
    private LocalDateTime editedAt;        // 编辑时间
    private LocalDateTime createdAt;       // 创建时间
    private LocalDateTime updatedAt;      // 更新时间
    private boolean deleted;               // 是否已删除（软删除）
    private int readCount;                // 已读人数
    private int unreadCount;               // 未读人数
    private int commentCount;              // 评论数

    public GroupAnnouncement() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
        this.pinned = false;
        this.edited = false;
        this.deleted = false;
        this.readCount = 0;
        this.unreadCount = 0;
        this.commentCount = 0;
    }

    // Getters and Setters
    public String getAnnouncementId() {
        return announcementId;
    }

    public void setAnnouncementId(String announcementId) {
        this.announcementId = announcementId;
    }

    public String getGroupId() {
        return groupId;
    }

    public void setGroupId(String groupId) {
        this.groupId = groupId;
    }

    public String getAuthorId() {
        return authorId;
    }

    public void setAuthorId(String authorId) {
        this.authorId = authorId;
    }

    public String getAuthorName() {
        return authorName;
    }

    public void setAuthorName(String authorName) {
        this.authorName = authorName;
    }

    public String getAuthorAvatar() {
        return authorAvatar;
    }

    public void setAuthorAvatar(String authorAvatar) {
        this.authorAvatar = authorAvatar;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getContent() {
        return content;
    }

    public void setContent(String content) {
        this.content = content;
        // 自动生成摘要
        if (content != null && content.length() > 100) {
            this.summary = content.substring(0, 100) + "...";
        } else {
            this.summary = content;
        }
    }

    public String getSummary() {
        return summary;
    }

    public void setSummary(String summary) {
        this.summary = summary;
    }

    public boolean isPinned() {
        return pinned;
    }

    public void setPinned(boolean pinned) {
        this.pinned = pinned;
    }

    public boolean isEdited() {
        return edited;
    }

    public void setEdited(boolean edited) {
        this.edited = edited;
    }

    public LocalDateTime getEditedAt() {
        return editedAt;
    }

    public void setEditedAt(LocalDateTime editedAt) {
        this.editedAt = editedAt;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }

    public boolean isDeleted() {
        return deleted;
    }

    public void setDeleted(boolean deleted) {
        this.deleted = deleted;
    }

    public int getReadCount() {
        return readCount;
    }

    public void setReadCount(int readCount) {
        this.readCount = readCount;
    }

    public int getUnreadCount() {
        return unreadCount;
    }

    public void setUnreadCount(int unreadCount) {
        this.unreadCount = unreadCount;
    }

    public int getCommentCount() {
        return commentCount;
    }

    public void setCommentCount(int commentCount) {
        this.commentCount = commentCount;
    }

    // Builder模式
    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {
        private final GroupAnnouncement announcement;

        public Builder() {
            this.announcement = new GroupAnnouncement();
        }

        public Builder announcementId(String announcementId) {
            announcement.setAnnouncementId(announcementId);
            return this;
        }

        public Builder groupId(String groupId) {
            announcement.setGroupId(groupId);
            return this;
        }

        public Builder authorId(String authorId) {
            announcement.setAuthorId(authorId);
            return this;
        }

        public Builder authorName(String authorName) {
            announcement.setAuthorName(authorName);
            return this;
        }

        public Builder authorAvatar(String authorAvatar) {
            announcement.setAuthorAvatar(authorAvatar);
            return this;
        }

        public Builder title(String title) {
            announcement.setTitle(title);
            return this;
        }

        public Builder content(String content) {
            announcement.setContent(content);
            return this;
        }

        public Builder summary(String summary) {
            announcement.setSummary(summary);
            return this;
        }

        public Builder pinned(boolean pinned) {
            announcement.setPinned(pinned);
            return this;
        }

        public Builder createdAt(LocalDateTime createdAt) {
            announcement.setCreatedAt(createdAt);
            return this;
        }

        public Builder updatedAt(LocalDateTime updatedAt) {
            announcement.setUpdatedAt(updatedAt);
            return this;
        }

        public GroupAnnouncement build() {
            return announcement;
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        GroupAnnouncement that = (GroupAnnouncement) o;
        return Objects.equals(announcementId, that.announcementId);
    }

    @Override
    public int hashCode() {
        return Objects.hash(announcementId);
    }

    @Override
    public String toString() {
        return "GroupAnnouncement{" +
                "announcementId='" + announcementId + '\'' +
                ", groupId='" + groupId + '\'' +
                ", authorId='" + authorId + '\'' +
                ", title='" + title + '\'' +
                ", pinned=" + pinned +
                ", deleted=" + deleted +
                '}';
    }
}
